src/transformers/modeling_reformer.py

```diff
@@ -0,0 +1,1760 @@
+# coding=utf-8
+# Copyright 2020 The Trax Authors and The HuggingFace Inc. team.
+# Copyright (c) 2018, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""PyTorch REFORMER model. """
+
+import logging
+import sys
+from collections import namedtuple
+from functools import reduce
+from operator import mul
+
+import numpy as np
+import torch
+from torch import nn
+from torch.autograd.function import Function
+from torch.nn import CrossEntropyLoss
+
+from .activations import gelu, gelu_fast, gelu_new, swish
+from .configuration_reformer import ReformerConfig
+from .file_utils import DUMMY_INPUTS, DUMMY_MASK, add_start_docstrings, add_start_docstrings_to_callable
+from .modeling_utils import PreTrainedModel, apply_chunking_to_forward
+
+
+logger = logging.getLogger(__name__)
+
+REFORMER_PRETRAINED_MODEL_ARCHIVE_MAP = {
+    "google/reformer-crime-and-punishment": "https://cdn.huggingface.co/google/reformer-crime-and-punishment/pytorch_model.bin"
+}
+
+
+def mish(x):
+    return x * torch.tanh(nn.functional.softplus(x))
+
+
+ACT2FN = {
+    "gelu": gelu,
+    "relu": torch.nn.functional.relu,
+    "swish": swish,
+    "gelu_new": gelu_new,
+    "gelu_fast": gelu_fast,
+    "mish": mish,
+}
+ReformerLayerNorm = torch.nn.LayerNorm
+
+
+# Define named tuples for nn.Modules here
+LSHSelfAttentionOutput = namedtuple("LSHSelfAttentionOutput", ["hidden_states", "attention_probs", "buckets"])
```

Comment:
@thomwolf @LysandreJik @julien-c I started using named tuples here. This is a proposal for a naming convention and how they could be used. All named tuples and with ``...Output`` and are defined at the top of the file. What do you think?

Reply:
Not yet because I wanted to ask first how to to handle this. The problem is that for named tuples "optional" outputs or tuples with "default" values are allowed only since Python 3.7, so no backward compatibility...Also I think it would actually be better to allows return *something* for `all_attentions` and `all_hidden_states` either an empty list / tuple or a full one. But maybe we should decide this in a different PR